### ISSUE TITLE
apiserver dies on long-running cluster: No connection could be made because the target machine actively refused it.

### ISSUE THREAD

**Issue by jgoeres** [kubernetes/minikube#8300]
I repeatedly have the situation that my minikube is up and running for a while. In particular after deleting and re-installing a slighly bigger workload with helm, kubectl would suddenly fail with 

    Unable to connect to the server: dial tcp 192.168.99.146:8443: connectex: No connection could be made because the target machine actively refused it.

I have this every now and then and so far I usually just deleted and restarted minikube. But today I ran into this 
https://github.com/kubernetes/minikube/issues/8298
and so I finally decided to open an issue for this.  ;-)

I fully understand that without any additional information, no one will be able to even analyze this, so please advise me where to look for logs etc. on the minikube VM when this happens again.

**Comment by tstromberg** [comment]
Hey @jgoeres - the things that would be useful are mentioned in the "New Issue" template:

* Steps to reproduce the issue, including the `minikube start` command used.
* The full output of `minikube start` command used
* The full output of `minikube logs` command:

The combination of these 3 should make it possible to debug this. In the past, when i've seen this issue, it's happened when the cluster ran out of memory, and the apiserver was sacrificed. This shouldn't happen with modern versions of minikube, however.

**Comment by jgoeres** [comment]
Okay, so it just happend again.
The minikube start command and its output:

    $ minikube start --disk-size 30G --cpus 6 --insecure-registry <OUR_REGISTRY> --kubernetes-version=v1.15.10 --memory 12g --nat-nic-type=Am79C973 --host-only-nic-type=Am79C973
    * minikube 1.10.1 is available! Download it: https://github.com/kubernetes/minikube/releases/tag/v1.10.1
    * To disable this notice, run: 'minikube config set WantUpdateNotification false'
    
    ! minikube v1.8.2 on Microsoft Windows 10 Enterprise 10.0.17763 Build 17763
    * Automatically selected the virtualbox driver
    * Creating virtualbox VM (CPUs=6, Memory=12000MB, Disk=30000MB) ...
    * Found network options:
      - NO_PROXY=192.168.99.100,192.168.99.140
      - no_proxy=192.168.99.100,192.168.99.140
    * Preparing Kubernetes v1.15.10 on Docker 19.03.6 ...
      - env NO_PROXY=192.168.99.100,192.168.99.140
      - env NO_PROXY=192.168.99.100,192.168.99.140
    * Launching Kubernetes ...
    * Enabling addons: default-storageclass, storage-provisioner
    * Waiting for cluster to come online ...
    * Done! kubectl is now configured to use "minikube"

The minkube logs command fails with

	$ minikube logs
	*
	X command runner
	* Error: [OOM_KILL_SCP] getting ssh client for bootstrapper: Error dialing tcp via ssh client: ssh: handshake failed: read tcp 127.0.0.1:65283->127.0.0.1:47677: wsarecv: An existing connection was forcibly closed by the remote
	 host.
	* Suggestion: Disable dynamic memory in your VM manager, or pass in a larger --memory value
	* Related issues:
	  - https://github.com/kubernetes/minikube/issues/1766

The issue actually refers to hyperv, while I am using virtualbox. I checked if there is anything in the VM settings that appears similar to the hyperv "dynamic memory" setting, but found nothing.
The apiserver container was running, BTW... actually there are multiple apiserver containers, one started, the others "created":

	$ docker ps -a | grep kube-ap
	98ca7c70cf75        af9a09211e3d            "kube-apiserver --ad…"   18 seconds ago      Created         k8s_kube-apiserver_kube-apiserver-m01_kube-system_b60b687ab7fcf5f4da64fb2930d1415e_18
	14f415faaab3        af9a09211e3d            "kube-apiserver --ad…"   5 minutes ago       Created         k8s_kube-apiserver_kube-apiserver-m01_kube-system_b60b687ab7fcf5f4da64fb2930d1415e_17
	7baecbe31cb7        af9a09211e3d            "kube-apiserver --ad…"   21 hours ago        Up 21 hours     k8s_kube-apiserver_kube-apiserver-m01_kube-system_b60b687ab7fcf5f4da64fb2930d1415e_0
	49ec73d9bbb6        k8s.gcr.io/pause:3.1    "/pause"                 21 hours ago        Up 21 hours     k8s_POD_kube-apiserver-m01_kube-system_b60b687ab7fcf5f4da64fb2930d1415e_0

**Comment by tstromberg** [comment]
@jgoeres - everything about this screams that this VM is running out of memory in a bad way: not only is the apiserver dead, but ssh is dead too. I'm quite honestly surprised that it was able to get this far.

It's difficult to debug a cluster if both sshd and apiserver are dying. Here's what I suggest:

* Upgrade to the latest version of minikube
* Run `minikube delete`
* Run `minikube start <your flags>`
* Run `minikube logs -f`

Run your workloads -- what I am hoping for is that when things begin falling apart, the `logs -f` command will emit useful enough messages to isolate which process is stealing all of the RAM from the usual Linux daemons.

**Comment by jgoeres** [comment]
THX, I will try this out and get back to you once it happens again, hopefully with more helpful logs.

**Comment by jgoeres** [comment]
So it happened again, find the (slightly anonymized) logs attached.
One more note: I had a watch running on a "kubectl get pods -A". The last thing I saw was that the elasticsearch that we are starting (custom image) was stuck in its init container (which does a "sysctl -w vm.max_map_count=262144").
This worked fine when I was using the ECK operator to set up elasticsearch (which uses the official ES image, and which also has an init container doing the same thing...).

[minikube_logs.txt](https://github.com/kubernetes/minikube/files/4715157/minikube_logs.txt)

**Comment by tstromberg** [comment]
I took a brief look, and it's apparent that the cluster is struggling for resources. I don't see any OOM messages, but I do see many etcd related delays. Did the `minikube logs` command ever die, or does it keep rolling?

Would it be possible to produce the manifests/commands used to generate a similar workload?

- The SSH error makes it sound like OOM
- `minikube logs` shows no signs of OOM, but some signs of CPU pressure

This leaves me a bit confused and unsure how to debug further, since we eventually lose ssh access seemingly.

**Comment by jgoeres** [comment]
Hi,

thx for the feedback. Alas, I don't think I can provide a way to deploy this workload for reproducing it. It mostly consists of images only available internally, used for hosting a SaaS. Basically a bunch (4-5) tomcats, one elasticsearch, one cassandra, one Zookeeper, and an Apache HTTPD.
Generally, CPU limitation causing the problem makes quite some sense - I assigned 6 cores to the minikube VM, although the dev machine has only 4+4 "cores" (not sure how virtual box handles the virtual "hyperthreading cores). But with fewer cores, I would run into problems during application startup, so assigning more than the actual physical cores and hoping for HT to do any good was 
successful to some degree. During startup, the machine's load factor often reaches ~20. But this worked quite OK initially. At some point, I started experiencing this issue - in particular when repeatedly discarding and reinstalling the workload. And that was already after we actually reduced the number of components in our stack. 
We will switch to a new gen of dev machines soon (6+6) but this has been delayed (guess why... ;-)).

In any case, there is probably little that can be done here, aside from killing it with iron. ;-)
